src/bsp/platform/esp32/xi_bsp_io_fs_esp32.c

```diff
@@ -0,0 +1,307 @@
+
+        *resource_handle_out = ( xi_bsp_io_fs_resource_handle_t )fp;
+
+        return XI_BSP_IO_FS_STATE_OK;
+    }
+
+    return XI_BSP_IO_FS_STATE_OK;
+
+err_handling:
+    return ret;
+}
+
+xi_bsp_io_fs_state_t
+xi_bsp_io_fs_write( const xi_bsp_io_fs_resource_handle_t resource_handle,
+                    const uint8_t* const buffer,
+                    const size_t buffer_size,
+                    const size_t offset,
+                    size_t* const bytes_written )
+{
+    ( void )offset;
+    xi_bsp_io_fs_state_t ret = XI_BSP_IO_FS_STATE_OK;
+    esp_err_t retv           = ESP_OK;
+
+    if ( ( esp_ota_handle_t )resource_handle == open_firmware_bin_handle )
+    {
+        retv = esp_ota_write( resource_handle, buffer, buffer_size );
+        if ( ESP_OK != retv )
+        {
+            xi_bsp_debug_format( "esp_ota_write() failed with error %d", retv );
+            return XI_BSP_IO_FS_WRITE_ERROR;
+        }
+
+        *bytes_written = buffer_size;
+        esp32_xibsp_notify_chunk_written( buffer_size, offset );
+    }
+    else
+    {
+        if ( NULL == buffer || 0 == buffer_size ||
+             XI_BSP_IO_FS_INVALID_RESOURCE_HANDLE == resource_handle )
+        {
+            return XI_BSP_IO_FS_INVALID_PARAMETER;
+        }
+
+        FILE* fp = ( FILE* )resource_handle;
+
+        /* let's set the offset */
+        const int fop_ret = fseek( fp, offset, SEEK_SET );
+
+        /* if error on fseek check errno */
+        XI_BSP_IO_FS_CHECK_CND(
+            fop_ret != 0, xi_bsp_io_fs_posix_errno_2_xi_bsp_io_fs_state( errno ), ret );
+
+        *bytes_written = fwrite( buffer, ( size_t )1, buffer_size, fp );
+
+        /* if error on fwrite check errno */
+        XI_BSP_IO_FS_CHECK_CND(
+            buffer_size != *bytes_written,
+            xi_bsp_io_fs_posix_errno_2_xi_bsp_io_fs_state( ferror( fp ) ), ret );
+    }
+    return XI_BSP_IO_FS_STATE_OK;
+
+err_handling:
+    return ret;
+}
+
+xi_bsp_io_fs_state_t
+xi_bsp_io_fs_read( const xi_bsp_io_fs_resource_handle_t resource_handle,
+                   const size_t offset,
+                   const uint8_t** buffer,
+                   size_t* const buffer_size )
+{
+    xi_bsp_io_fs_state_t ret = XI_BSP_IO_FS_STATE_OK;
+    FILE* fp                 = ( FILE* )resource_handle;
+    int fop_ret              = 0;
+
+    if ( XI_BSP_IO_FS_INVALID_RESOURCE_HANDLE == resource_handle )
+    {
+        xi_bsp_debug_logger( "\nInvalid resource handle - Filesystem read failed" );
+        return XI_BSP_IO_FS_INVALID_PARAMETER;
+    }
+
+    if ( ( esp_ota_handle_t )resource_handle == open_firmware_bin_handle )
+    {
+        xi_bsp_debug_logger( "Error: No FS read of FW files in ESP32's OTA API" );
+        return XI_BSP_IO_FS_NOT_IMPLEMENTED;
+    }
+    else
+    {
+        /* make an allocation for memory block */
+        if ( NULL == *buffer )
+        {
+            *buffer = ( uint8_t* )xi_bsp_mem_alloc( xi_bsp_io_fs_buffer_size );
```

Comment:
this memory chunk is never deallocated: leaks

Reply:
please check the xi_bsp_io_fs_cc3200.c's read function and the buffer used there. A local BSP IO FS static buffer is used there and returned. Since other BSPs are written so, libxively internals (in this case xi_sft_revision.c) assume no deallocation is needed. So just define here a local static read buffer array, like this:
static uint8_t read_buffer[ xi_bsp_io_fs_buffer_size];

Reply:
☠️ ☠️☠️

Nice catch, thanks for the heads-up. Fixed 👍